apps/pipelines/nodes/history_middleware.py

```diff
@@ -0,0 +1,157 @@
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING
+
+from langchain.agents.middleware.summarization import SummarizationMiddleware
+from langchain_core.messages import BaseMessage, RemoveMessage
+from langchain_core.messages.utils import count_tokens_approximately
+from langgraph.graph.message import (
+    REMOVE_ALL_MESSAGES,
+)
+
+from apps.chat.conversation import COMPRESSION_MARKER
+from apps.chat.models import ChatMessage
+from apps.pipelines.models import PipelineChatHistoryModes, PipelineChatMessages
+
+if TYPE_CHECKING:
+    from apps.pipelines.nodes.nodes import PipelineNode
+
+logger = logging.getLogger("ocs.bots.summarization")
+
+
+class HistoryCompressionMiddleware(SummarizationMiddleware):
+    """
+    Middleware to summarize chat history based on node configuration.
+
+    History is loaded into the state in the before_agent step, and summaries are persisted
+    to the database in the before_model step, if a summary is generated.
+    """
+
+    def __init__(self, session, node: PipelineNode, **kwargs):
+        super().__init__(model=node.get_chat_model(), **kwargs)
+        self.session = session
+        self.node = node
+
+    def before_agent(self, state, runtime):
+        return {
+            "messages": [
+                # Since this response will get merged with the existing state messages, we cannot simply append the
+                # history to the user's message. We need to replace the full message history in the state.
+                # See https://github.com/langchain-ai/langchain/blob/c63f23d2339b2604edc9ae1d9f7faf7d6cc7dc78/libs/langchain_v1/langchain/agents/middleware/summarization.py#L286-L292
+                RemoveMessage(id=REMOVE_ALL_MESSAGES),
+                *self.node.get_history(self.session),
+                *state["messages"],
+            ]
+        }
+
+    def before_model(self, state, runtime):
+        result = super().before_model(state, runtime)
+        if result is not None:
+            self._persist_summary(result["messages"])
+
+        return result
+
+    def _create_summary(self, messages_to_summarize: list) -> str:
+        """
+        Create summary based on node's history mode
+        TRUNCATE_TOKENS and MAX_HISTORY_LENGTH mode does not create summaries, it only removes old messages until under
+        token or message limit.
+        """
+        if self.node.history_mode == PipelineChatHistoryModes.TRUNCATE_TOKENS:
+            return COMPRESSION_MARKER
```

Comment:
@snopoke I don't see the value in the compression marker other than the fact that it will prevent loading the whole history when we call `get_langchain_messages_until_marker`. Am I missing some other usecase?

Reply:
So I'm assuming we'll want to keep this around then